### PR TITLE
add doc for jenkins login via openshift oauth

### PR DESCRIPTION
**Pull request by gabemontero**
@bparees ptal (but wait until next week unless the weather is so bad where you are vacationing that you are stuck in your hotel room bored :-) )

### CODE REVIEW

File: using_images/other_images/jenkins.adoc
Comment (bparees) [inline using_images/other_images/jenkins.adoc]: doesn't this boil down to "jenkins auth will be used by default, openshift oauth will be used when the image is running in a pod on an openshift cluster and the openshift_enable_oauth env variable is set to a value other than 'false' "?

also this implies we can't enable openshift oauth when running outside openshift...is that true? if so why?

File: using_images/other_images/jenkins.adoc
Comment (bparees) [inline using_images/other_images/jenkins.adoc]: why is this ifdef'd to origin? it applies to OCP too right?  (starting with v3.4 so you'll need to label the PR appropriately)

File: using_images/other_images/jenkins.adoc
Comment (bparees) [inline using_images/other_images/jenkins.adoc]: i think it makes more sense to say "default user/login is blah when openshift oauth authentication is not enabled"

File: using_images/other_images/jenkins.adoc
Comment (bparees) [inline using_images/other_images/jenkins.adoc]: this isn't accurate.  the randomly password behavior only applies when instantiating the jenkins template.  this is just using new-app to stand up the image directly, not via the template, so the parameter generation logic is not applicable.

File: using_images/other_images/jenkins.adoc
Comment (bparees) [inline using_images/other_images/jenkins.adoc]: again i don't think this should be if-def'd

File: using_images/other_images/jenkins.adoc
Comment (bparees) [inline using_images/other_images/jenkins.adoc]: s/perhaps//

File: using_images/other_images/jenkins.adoc
Comment (bparees) [inline using_images/other_images/jenkins.adoc]: s/you want to access/accessing/

File: using_images/other_images/jenkins.adoc
Comment (bparees) [inline using_images/other_images/jenkins.adoc]: we should be encouraging users to use their openshift username to login, not "admin" since "admin" is only going to work when the cluster has the Allow All config (or the user's openshift name really is "admin")

File: using_images/other_images/jenkins.adoc
Comment (bparees) [inline using_images/other_images/jenkins.adoc]: need clarification around what happens when you change permissions in jenkins and then polling happens (and either the openshift permissions have changed, or have not changed but conflict with the changes made in jenkins)

File: using_images/other_images/jenkins.adoc
Comment (bparees) [inline using_images/other_images/jenkins.adoc]: no ifdef

File: using_images/other_images/jenkins.adoc
Comment (bparees) [inline using_images/other_images/jenkins.adoc]: "Enabled when set to any non-empty value other than "false""

File: using_images/other_images/jenkins.adoc
Comment (bparees) [inline using_images/other_images/jenkins.adoc]: i guess we didn't learn our lesson about defaulting to milliseconds on this stuff?  :)

can we change this to seconds please?

**Review by bparees — CHANGES_REQUESTED** [review]

File: using_images/other_images/jenkins.adoc
Comment (gabemontero) [inline using_images/other_images/jenkins.adoc]: I ifdef'ed cause I thought that is what was required until 3.4 GA'ed.  I am not allowed to add labels to the PR, but will remove the ifdef.

File: using_images/other_images/jenkins.adoc
Comment (gabemontero) [inline using_images/other_images/jenkins.adoc]: As part of getting rid of the ifdef's, I'm getting rid of any mention of oauth in the default jenkins auth section.  The openshift oauth section explains the default means for logging in for that config.

File: using_images/other_images/jenkins.adoc
Comment (gabemontero) [inline using_images/other_images/jenkins.adoc]: ok i'll just remove

File: using_images/other_images/jenkins.adoc
Comment (gabemontero) [inline using_images/other_images/jenkins.adoc]: ok ... i'll have to revisit one aspect of my testing before i agree - with the requirement of using the self-sar to get permissions, it thought the user that created the project did not have the admin role with the OOTB `Allow All` scenario; if correct, then I think this impacts the OOTB experience, as the developer won't have enough power in Jenkins to do stuff.

I'll circle back and report findings once I get to try it, and we'll go from there.

File: using_images/other_images/jenkins.adoc
Comment (gabemontero) [inline using_images/other_images/jenkins.adoc]: this section was the same as the jenkins repo readme ... pulled my clarification there here.

File: using_images/other_images/jenkins.adoc
Comment (gabemontero) [inline using_images/other_images/jenkins.adoc]: OK, I was able to confirm that the user ID that created the project does in fact have the admin role in the OOTB scenario.

I suppose it is safe to assume that if say a user with OpenShift Online instantiates the jenkins template in his project, the same will apply.

I'll rework the document to encourage the use of the user that he logged into openshift with / created the project under.

**Comment by gabemontero** [comment]
OK addressed the use of `admin` as the user name.  Also moved some of the more esoteric options that were in the origin READMEs into the product doc.

File: using_images/other_images/jenkins.adoc
Comment (bparees) [inline using_images/other_images/jenkins.adoc]: s/the the/the/

File: using_images/other_images/jenkins.adoc
Comment (bparees) [inline using_images/other_images/jenkins.adoc]: it won't be a randomly generated string unless they instantiate it from a template.  if they just run the image w/o setting JENKINS_PASSWORD (which is what the template does), it'll still be "password", right?

File: using_images/other_images/jenkins.adoc
Comment (bparees) [inline using_images/other_images/jenkins.adoc]: "you will have administrative privileges within Jenkins."

File: using_images/other_images/jenkins.adoc
Comment (bparees) [inline using_images/other_images/jenkins.adoc]: this is based on the role of the user within the project that is hosting jenkins.  What's the behavior when jenkins is run outside of openshift but still uses oauth?

File: using_images/other_images/jenkins.adoc
Comment (bparees) [inline using_images/other_images/jenkins.adoc]: change sentence starting w/ "Technically speaking" to:

"If the Jenkins UI is used to update permissions for a Jenkins user, the permission changes will be overwritten the next time the plugin polls {product-title}."

File: using_images/other_images/jenkins.adoc
Comment (bparees) [inline using_images/other_images/jenkins.adoc]: "The token field contains the token value Jenkins will need to access the project."

File: using_images/other_images/jenkins.adoc
Comment (bparees) [inline using_images/other_images/jenkins.adoc]: "Identify the service account token secret associated with a service account that has appropriate permissions to access the project Jenkins will need to access:"

**Review by bparees — CHANGES_REQUESTED** [review]
couple more changes and i think it's good.

File: using_images/other_images/jenkins.adoc
Comment (gabemontero) [inline using_images/other_images/jenkins.adoc]: This gets into the scenario Clayton wasn't interested in documenting when I asked him (you were either traveling from Brno or on PTO when I pinged him)- that it just needed to be possible for a developer to hack it.  You have to explicitly configure the token, cert, namespace, client id, redirect url, and api master url.  And in the case of the namespace and the cert, I have not improved upon what Clayton put together / pulled from the google version of this, and those items have to be mounted in from a file system when you `docker run ...` the image outside of OpenShift.

Do you want to resurface this discussion / put more spit and polish on the outside OpenShift use case.  Or minimally put some disclaimer or something here?

**Comment by gabemontero** [comment]
OK I've pushed the minor wording updates and clarifications.  Pending validation of those for correctness, there is the "what to do about running oauth when Jenkins is outside of OpenShift" question remaining.

File: using_images/other_images/jenkins.adoc
Comment (bparees) [inline using_images/other_images/jenkins.adoc]: Nope, that's fine, mostly i just wanted to know for myself.  But be aware that someone is going to ask and you're going to have to explain it to them, so it'd be to your benefit if it's written down somewhere.  Maybe as part of the oauth plugin repo's readme/doc?

File: using_images/other_images/jenkins.adoc
Comment (bparees) [inline using_images/other_images/jenkins.adoc]: this new section is inserted right in the middle of the lead in to the "new-app" command for launching the image directly(no template), and the actual example new-app command that launches the image(no template).  I think it needs restructuring.  Also if they use the template, then they get oauth behavior, right?  so do we even need to talk about this randomly generated password scenario anymore?

**Review by bparees — CHANGES_REQUESTED** [review]

**Comment by gabemontero** [comment]
That is a good idea (repo readme)

I'll put the gorpy detail there.

File: using_images/other_images/jenkins.adoc
Comment (gabemontero) [inline using_images/other_images/jenkins.adoc]: If they instantiate the template with the env prop set to something other than false, they get oauth.  It is fairly straight forward to get the standard jenkins authentication enabled.  I think we still need coverage for this.

I'll take a second pass at the placement of the section and push something up in a bit.

Comment (gabemontero) [inline using_images/other_images/jenkins.adoc]: OK, I've pushed the corresponding update.  I reordered the presentation of the auth methods, putting oauth first.  I also added `oc new-app` examples for both ensuring oauth is the login method and ensuring standard jenkins is the standard method.  I think it flows better - but let me know.

File: using_images/other_images/jenkins.adoc
Comment (bparees) [inline using_images/other_images/jenkins.adoc]: they don't actually need to set this, right?  it's the default behavior.

File: using_images/other_images/jenkins.adoc
Comment (bparees) [inline using_images/other_images/jenkins.adoc]: this still reads weirdly for me.  If they use the template, they're going to get oauth behavior(right? because we're going to craft the template to ensure that), so the randomly generated password is moot.  In fact we should remove the password parameter from the template when we update it to enable oauth.

I think if we are going to talk about standing up jenkins via the template at all, we should talk about it in the oauth section and not the standard auth section.

**Review by bparees — CHANGES_REQUESTED** [review]

File: using_images/other_images/jenkins.adoc
Comment (bparees) [inline using_images/other_images/jenkins.adoc]: no.

JENKINS_PASSWORD is still a config variable that is respected by the jenkins image (and this is the jenkins image documentation).  But we need to make it clear that the variable only matters when you're not using oauth, and if you instantiate via the template, you're using oauth so it's irrelevant there.

Comment (gabemontero) [inline using_images/other_images/jenkins.adoc]: Gotcha ... and running new-app against the image (openshift/jenkins-1-centos) like in the example is not instantiating from the template.

Based on that, I'll update the `oc new-app` example to reference the template and update the image.

Pushing momentarily.

Comment (gabemontero) [inline using_images/other_images/jenkins.adoc]: OK pushed update that has command invocations around instantiating templates in the oauth section, and and the oc new-app against the image for the standard auth.  Added clarifications the JENKINS_PASSWORD is only applicable to standard jenkins auth.

File: using_images/other_images/jenkins.adoc
Comment (bparees) [inline using_images/other_images/jenkins.adoc]: s/3.4/v3.4/

"standard Jenkins authentication will be used.  The default 'admin' account will exist with password 'password'"

Comment (bparees) [inline using_images/other_images/jenkins.adoc]: "See Jenkins standard authentication for details about changing this password."

File: using_images/other_images/jenkins.adoc
Comment (bparees) [inline using_images/other_images/jenkins.adoc]: i can't comment on it in this PR but line 171 above should be reworded to:  "The Jenkins server can be configured with the following environment variables:"

File: using_images/other_images/jenkins.adoc
Comment (bparees) [inline using_images/other_images/jenkins.adoc]: this should be jenkins-ephemeral-template.json

File: using_images/other_images/jenkins.adoc
Comment (bparees) [inline using_images/other_images/jenkins.adoc]: for both of these sets of commands we can assume the default imagestreams and templates are registered, don't make them run those commands again.  Generally we preface it with something like "Assuming you have installed the default imagestreams and templates (link to https://docs.openshift.org/latest/install_config/imagestreams_templates.html), you can run:"

**Review by bparees — CHANGES_REQUESTED** [review]

File: using_images/other_images/jenkins.adoc
Comment (gabemontero) [inline using_images/other_images/jenkins.adoc]: part of next push

File: using_images/other_images/jenkins.adoc
Comment (gabemontero) [inline using_images/other_images/jenkins.adoc]: part of next push

**Comment by gabemontero** [comment]
latest responses to comments pushed

**Comment by bparees** [comment]
ok i think i've got nothing else (i can hear your sigh of relief from here).  probably hold off on tagging the docs team until all the actual functional stuff has merged though, in case anything else changes.

**Comment by gabemontero** [comment]
On Tue, Oct 11, 2016 at 3:47 PM, Ben Parees <notifications@github.com>
wrote:

> ok i think i've got nothing else (i can hear your sigh of relief from
> here). probably hold off on tagging the docs team until all the actual
> functional stuff has merged though, in case anything else changes.
>
lol ... and yep, both this and the origin PR are on hold until the 2 PM PRs
are merged and we can cross the i's and dot the t's

**Comment by gabemontero** [comment]
@openshift/team-documentation PTAL

**Comment by vikram-redhat** [comment]
@bfallonf - PTAL.

**Comment by bfallonf** [comment]
@vikram-redhat @gabemontero @bparees I can merge this and do a followup PR, but I think Ben has to approve Gabe's changes before I can merge? I'm not sure how this fancy 'review' thing works, so I don't want to 'dismiss' Ben's review.

**Comment by bparees** [comment]
I gave approval in my comment earlier in the thread.

**Review by bparees — APPROVED** [review]

**Comment by bfallonf** [comment]
@bparees Thanks! Just wanted the button to be pressed. I'll merge away.

**Comment by gabemontero** [comment]
Hey - Ben gave the "igtm" previously, but sure we can wait and let him
chime in.

**Comment by bfallonf** [comment]
@gabemontero Yup, he gave us the thumbs up. It was more about the "requested changes" review being given the go-ahead. Thanks, all!

**Pull request merged**
